### ISSUE TITLE
Backup shows "Invalid Date"

### ISSUE THREAD

**Issue by ainiml** [longhorn/longhorn#987]
Backup data is still intact, and exists.

Browser: `Ubuntu Chrome Version 81.0.4000.3 (Official Build) dev (64-bit)`

![image](https://user-images.githubusercontent.com/28725423/71539147-270e1b80-2906-11ea-975b-4c4282068620.png)

Possibly related: https://github.com/longhorn/longhorn/issues/227

**Comment by ainiml** [comment]
Is there a way to fix this manually, so I can continue testing? Otherwise, I'll have to delete everything and re-create all the volumes

**Comment by yasker** [comment]
@smallteeths What's the reason for `CreatedAt` to show `Invalid date`?

@ainiml What's the result if you use `<longhorn_url>/v1/backupvolumes`? This is the API UI was reading from, can you check the field `created` here?

**Comment by ainiml** [comment]
> @smallteeths What's the reason for `CreatedAt` to show `Invalid date`?
> 
> @ainiml What's the result if you use `<longhorn_url>/v1/backupvolumes`? This is the API UI was reading from, can you check the field `created` here?

Is the URL like this? `https://longhorn:8443/k8s/clusters/c-jzh49/api/v1/backupvolumes`

```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "the server could not find the requested resource",
  "reason": "NotFound",
  "details": {
    
  },
  "code": 404
}
```

**Comment by yasker** [comment]
Oh, our management API URL doesn't work behind the Rancher proxy currently...

@ainiml Is it possible for you to create a node-port or xip ingress temporarily to access the management API? You can create a node port service in `longhorn-system` with target set to `longhorn-ui` and port `8000`. Remember to remove the service later though.

**Comment by ainiml** [comment]
> Oh, our management API URL doesn't work behind the Rancher proxy currently...
> 
> @ainiml Is it possible for you to create a node-port or xip ingress temporarily to access the management API? You can create a node port service in `longhorn-system` with target set to `longhorn-ui` and port `8000`. Remember to remove the service later though.

volume.cfg indeed does not exist in that path:

`"error": "cannot find backupstore/volumes/83/72/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66/volume.cfg in backupstore",`

![image](https://user-images.githubusercontent.com/28725423/71687043-d9147000-2d6a-11ea-8f38-9ac76cb5656e.png)


```
"actions": {
"backupDelete": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66?action=backupDelete",
"backupGet": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66?action=backupGet",
"backupList": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66?action=backupList",
},
"backups": null,
"baseImage": "",
"created": "",
"dataStored": "0",
"id": "pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66",
"lastBackupAt": "",
"lastBackupName": "",
"links": {
"self": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66",
},
"messages": {
"error": "cannot find backupstore/volumes/83/72/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66/volume.cfg in backupstore",
},
"name": "pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66",
"size": "0",
"type": "backupVolume",
},
```

But there are plenty of backup configs

![image](https://user-images.githubusercontent.com/28725423/71687127-1547d080-2d6b-11ea-91b1-d36744d080c3.png)


I'll try to restore `volume.cfg` from one of the backups

**Comment by ainiml** [comment]
> Oh, our management API URL doesn't work behind the Rancher proxy currently...
> 
> @ainiml Is it possible for you to create a node-port or xip ingress temporarily to access the management API? You can create a node port service in `longhorn-system` with target set to `longhorn-ui` and port `8000`. Remember to remove the service later though.

volume.cfg indeed does not exist in that path:

`"error": "cannot find backupstore/volumes/83/72/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66/volume.cfg in backupstore",`

![image](https://user-images.githubusercontent.com/28725423/71687043-d9147000-2d6a-11ea-8f38-9ac76cb5656e.png)


```
"actions": {
"backupDelete": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66?action=backupDelete",
"backupGet": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66?action=backupGet",
"backupList": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66?action=backupList",
},
"backups": null,
"baseImage": "",
"created": "",
"dataStored": "0",
"id": "pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66",
"lastBackupAt": "",
"lastBackupName": "",
"links": {
"self": "…/v1/backupvolumes/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66",
},
"messages": {
"error": "cannot find backupstore/volumes/83/72/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66/volume.cfg in backupstore",
},
"name": "pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66",
"size": "0",
"type": "backupVolume",
},
```

But there are plenty of backup configs

![image](https://user-images.githubusercontent.com/28725423/71687127-1547d080-2d6b-11ea-91b1-d36744d080c3.png)


I'll try to restore `volume.cfg` from one of the backups

**Comment by yasker** [comment]
That's weird. Can you check if you make a new backup, would the `volume.cfg` be there? Maybe something happened to the `volume.cfg`.

**Comment by ainiml** [comment]
@yasker 

Copying over the backup.cfg allowed you to see the backups again


```
~ # mc cp dream/longhorn/backupstore/volumes/83/72/pvc-ebe9fe2b-787a-4166-b55a-3b89d768ea66/backups/backup_backup-182693ffe65842d1.cfg dream/longhorn/backupstore/volumes/83/72/pvc-ebe9fe2b-787a-4166-b55a-3b89d76
8ea66/volume.cfg
...ckup_backup-182693ffe65842d1.cfg:  3.83 KiB / 3.83 KiB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 14.27 KiB/s 0s~ # 
```

![image](https://user-images.githubusercontent.com/28725423/71688335-ac158c80-2d6d-11ea-880d-0ee21492773d.png)

**Comment by ainiml** [comment]
@yasker 

So something happened to the `volume.cfg` file, but it has backup configs. Is it possible for Longhorn to restore from the backup?

Since the name and path to the volume and backups are pretty unique, I think it could be safe to restore what's missing automatically (with some sort of initial error)?

**Comment by ainiml** [comment]
@yasker 

I think it might also be because the backup backend doesn't use something like copy-on-write, so if writing to the file is interrupted, it ends up being corrupted, and deleted (might be why `volume.cfg` doesn't exist)

**Comment by ainiml** [comment]
@yasker 

I just copied all the backups back to `volume.cfg`, and the backups are showing again. That took quite a lot of time to do

![image](https://user-images.githubusercontent.com/28725423/71689362-dec08480-2d6f-11ea-9de0-f5a41a8b04fb.png)

**Comment by yasker** [comment]
@ainiml It's not expected. We want to check more on how can this happen.

**Comment by ainiml** [comment]
> @ainiml It's not expected. We want to check more on how can this happen.

My best guess is the backend crashes. And when scrubbing on repair and remount, it deletes the corrupted files, including `volume.cfg`

**Comment by yasker** [comment]
@ainiml Do you mean backup target is NFS server on the same node? We recommend creating backups outside the Kubernetes cluster. Otherwise, if you lost the cluster, you can lose all your data and backups.

**Comment by ainiml** [comment]
> @ainiml Do you mean backup target is NFS server on the same node? We recommend creating backups outside the Kubernetes cluster. Otherwise, if you lost the cluster, you can lose all your data and backups.

Backup target is a Minio bucket. Minio backend is s3ql. S3ql backend is S3.

Longhorn backend is XFS on s3backer. S3backer backend is S3.

**Comment by yasker** [comment]
It's hard for us to support anything other than bare metal disk or cloud provider provided disks. I think s3backer is the problem here.

**Comment by ainiml** [comment]
> It's hard for us to support anything other than bare metal disk or cloud provider provided disks. I think s3backer is the problem here.

yeah s3backer crashes when restoring all the volumes all at once. We're restoring the backups individually now.

**Comment by derekbit** [comment]
Closed because the user hasn't responded for a long time.
